### PR TITLE
PPC 2.0: Fix accessibility issues in the new checklist

### LINKED ISSUE #8211: Editor: fix axe tests for checklist
## Bug Description

The axe tests are currently skipped because the page preview used to show the page components are using the same ids as other page previews in the app.

Make sure that all ids are unique. The tests are located here: https://github.com/google/web-stories-wp/pull/8178/files#diff-d9b4953bce7017875dbafc806140cf4d42a8761da3ac018f9e7c2bc7913ce8a6R301-R320

## Steps to Reproduce

Run the axe tests. You can use `fit` on each test to make them run alone. This is the command I was using

```bash
> npm run test:karma:edit-story:watch
```

## Screenshots

image

## Additional Context

The axe plugin is very helpful for determining accessibility problems: https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by samwhale**
## Context

PPC 2.0

## Summary

FIxes many accessibility issues in the checklist. Some were tracked in tickets and others were not.

## Relevant Technical Choices

n/a

## To-do

n/a

## User-facing changes

None

## Testing Instructions

The checklist should still work the same but now should be even more accessible than it was before.

this plugin was useful in identifying and fixing the accessibility violations: https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US. Thanks @BrittanyIRL 🥳 

To test:

**Easy way**: add that plugin to chrome and run the automatic scan when the checklist is open and closed. Should see no violations that are in the checklist (there are some but that's elsewhere).

**Hard way**: I added reasons for the changes in comments in the PR. Please review each change and use a screenreader or the devtools to verify the changes have been propagated.

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8211

### CODE REVIEW

File: assets/src/edit-story/components/carousel/carouselPage.js [hunk 1]
```diff
@@ -72,7 +72,7 @@ const PageSeparator = styled(ReorderableSeparator)`
   }
 `;
 
-const ReorderablePage = styled(ReorderableItem).attrs({ role: 'option' })`
+const ReorderablePage = styled(ReorderableItem)`
```
Comment (samwhale) [inline assets/src/edit-story/components/carousel/carouselPage.js:75]: `role="option"` was nested inside of another div with `role="option"`

File: assets/src/edit-story/components/checklist/checklist.js [hunk 2]
```diff
@@ -148,21 +148,21 @@ export function Checklist() {
...
-                isOpen={openPanel === ISSUE_TYPES.DESIGN}
+                isOpen={isOpen && openPanel === ISSUE_TYPES.DESIGN}
                 onClick={handleOpenPanel(ISSUE_TYPES.DESIGN)}
                 maxHeight={maxPanelHeight}
                 title={CATEGORY_LABELS[ISSUE_TYPES.DESIGN]}
               />
               <AccessibilityChecks
                 badgeCount={accessibilityBadgeCount}
-                isOpen={openPanel === ISSUE_TYPES.ACCESSIBILITY}
+                isOpen={isOpen && openPanel === ISSUE_TYPES.ACCESSIBILITY}
```
Comment (samwhale) [inline assets/src/edit-story/components/checklist/checklist.js:165]: Panels inside the tablist thought they were open when the popup wasn't open. This was leading to an axe violation where the scrollable content needed to be focusable when the checklist was closed.

File: assets/src/edit-story/components/checklistCard/checklistCard.js [hunk 3]
```diff
@@ -81,7 +81,7 @@ const ChecklistCard = ({
         <Title as={titleProps?.onClick ? 'button' : 'div'} {...titleProps}>
           <Headline
             size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.XXX_SMALL}
-            as="h4"
+            as="h2"
```
Comment (samwhale) [inline assets/src/edit-story/components/checklistCard/checklistCard.js]: heading violation for accessibility stuff. Needs to decrease by 1

File: assets/src/edit-story/components/inspector/prepublish/components/autoVideoOptimization.js [hunk 4]
```diff
@@ -69,14 +69,14 @@ const AutoVideoOptimization = ({
...
-          id="automatic-video-optimization-toggle"
+          id="auto-video-optimization-toggle"
           aria-label={__('Enable automatic video optimization', 'web-stories')}
           checked={areVideosAutoOptimized || false}
           onChange={onAutoOptimizeVideoClick}
         />
         <DescriptionText
           forwardedAs="label"
-          htmlFor="automatic-video-optimization-toggle"
+          htmlFor="auto-video-optimization-toggle"
```
Comment (samwhale) [inline assets/src/edit-story/components/inspector/prepublish/components/autoVideoOptimization.js:79]: These ids were the same as the ids in the old checklist. This won't matter once it's removed.

File: assets/src/edit-story/components/tablist/styles.js [hunk 5]
```diff
@@ -167,14 +168,19 @@ export const Badge = styled.div`
...
 export const TabPanel = styled.div`
   height: 0;
   padding: 0;
   visibility: hidden;
-  overflow-y: scroll;
+  overflow-y: hidden;
   overflow-x: hidden;
   background: ${({ theme }) => theme.colors.bg.primary};
-  transition: 250ms ease-in;
+  transition: height 250ms ease-in;
```
Comment (samwhale) [inline assets/src/edit-story/components/tablist/styles.js]: Cleanup and making sure the content in the panels is not picked up by screen readers when it's not supposed to be.

File: assets/src/edit-story/components/tablist/tablistPanel.js [hunk 6]
```diff
@@ -81,11 +76,15 @@ const TablistPanel = ({
           <PanelText aria-hidden>{badgeCount}</PanelText>
         </Badge>
       </TabButton>
-      <Scrollable maxHeight={maxHeight}>
-        <TabPanel aria-labelledby={`${title}-${panelId}`} role="tabpanel">
+      <ScrollableContent maxHeight={maxHeight}>
+        <TabPanel
+          id={panelId}
```
Comment (samwhale) [inline assets/src/edit-story/components/tablist/tablistPanel.js:88]: Needed this to make sure the button was identifying the content it's controlling with `aria-controls`. It's hidden just above this in the diff 👍

File: assets/src/edit-story/masks/display.js [hunk 7]
```diff
@@ -46,6 +48,9 @@ export default function WithMask({
   previewMode = false,
   ...rest
 }) {
+  // This component is used twice - random id appended to make sure
+  // id is unique for the Mask.
+  const randomId = useMemo(uuidv4, []);
```
Comment (samwhale) [inline assets/src/edit-story/masks/display.js:53]: We're using this component in the checklist now too. The ids were matching so we needed to make sure they were a little more random.

File: assets/src/edit-story/components/checklistCard/checklistCard.js (hunk 3 above)
Comment (BrittanyIRL) [reply assets/src/edit-story/components/checklistCard/checklistCard.js]: so i actually think this needs to be h4 and that the tab panel titles (priority, design, accessibility) need to be h3 and the popup itself needs to be h2. 

Screen Shot 2021-07-06 at 3 43 11 PM

this creates the following outline: 
- web stories
  - Checklist 
    - Priority Issues 
      - Add Web Story title 
      - Add Web story Description 

I use a `headingsMap` extension in chrome to help do sanity checks on this. The headings outline on the editor needs a little love, but this way the checklist isn't adding to that for later. 

The below is how it is on this branch, which you're right there was a violation in stepping from h1 immediately to h4. But the checks listed should be nested. It's like having an outline for a term paper - `Add web story title` isn't given much context about what it means by just `web stories`. 

Screen Shot 2021-07-06 at 3 45 50 PM

File: assets/src/edit-story/masks/display.js [hunk 8]
```diff
@@ -78,7 +83,7 @@ export default function WithMask({
 
   const maskId = `mask-${mask.type}-${element.id}-display${
     previewMode ? '-preview' : ''
-  }`;
+  }-${randomId}`;
```
Comment (BrittanyIRL) [inline assets/src/edit-story/masks/display.js:86]: 🎉

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
Thanks for taking the initiative to fix this! 

The only thing I see that should be changed before merging this is the headings.

File: assets/src/edit-story/components/checklistCard/checklistCard.js (hunk 3 above)
Comment (samwhale) [reply assets/src/edit-story/components/checklistCard/checklistCard.js]: Fixed it!

File: assets/src/edit-story/components/checklist/styles.js [hunk 9]
```diff
@@ -45,6 +45,10 @@ export const getTabPanelMaxHeight = (buttonCount) => {
 export const StyledTablistPanel = styled(TablistPanel)`
   height: ${({ badgeCount }) => (badgeCount === 0 ? 0 : 'auto')};
   visibility: ${({ badgeCount }) => (badgeCount === 0 ? 'hidden' : 'visible')};
+
+  button {
+    display: ${({ badgeCount }) => (badgeCount === 0 ? 'none' : 'auto')};
+  }
```
Comment (samwhale) [inline assets/src/edit-story/components/checklist/styles.js:51]: Button needed to be hidden when the panel is being hidden.

File: assets/src/edit-story/components/tablist/styles.js [hunk 10]
```diff
@@ -167,14 +168,19 @@ export const Badge = styled.div`
   }
 `;
 
+export const ScrollableContent = styled.div`
+  max-height: ${({ maxHeight }) => (maxHeight ? `calc(${maxHeight})` : 'none')};
+  overflow-y: ${({ maxHeight }) => (maxHeight ? 'scroll' : 'hidden')};
+`;
```
Comment (samwhale) [inline assets/src/edit-story/components/tablist/styles.js:174]: only moved

File: assets/src/edit-story/components/tablist/styles.js [hunk 11]
```diff
@@ -167,14 +168,19 @@ export const Badge = styled.div`
...
 export const TabPanel = styled.div`
   height: 0;
   padding: 0;
   visibility: hidden;
-  overflow-y: scroll;
+  overflow-y: hidden;
   overflow-x: hidden;
   background: ${({ theme }) => theme.colors.bg.primary};
-  transition: 250ms ease-in;
+  transition: all 300ms ease-in;
```
Comment (samwhale) [inline assets/src/edit-story/components/tablist/styles.js:183]: Fixes the animation grossness! Adds the `overflow-y: scroll` when the tab is expanded.

File: assets/src/edit-story/components/tablist/styles.js [hunk 12]
```diff
@@ -93,9 +93,9 @@ export const PanelWrapper = styled.div`
   width: 100%;
   background: ${({ theme }) => theme.colors.bg.primary};
 
-  :not(:last-child) ${TabButton} {
-    box-shadow: 0px 1px 0 0 ${({ theme }) => theme.colors.divider.tertiary};
-    margin-bottom: 1px;
+  &:not(:first-child) > ${TabButton} {
+    box-shadow: 0px -1px 0 0 ${({ theme }) => theme.colors.divider.tertiary};
+    margin-top: 1px;
```
Comment (samwhale) [inline assets/src/edit-story/components/tablist/styles.js:98]: fixes shadow styles when buttons are missing

File: assets/src/edit-story/components/tablist/tablistPanel.js [hunk 13]
```diff
@@ -73,19 +62,36 @@ const TablistPanel = ({
           <IconContainer>
             <Icons.ChevronDownSmall />
           </IconContainer>
-          <PanelText id={`${title}-${panelId}`} aria-hidden>
+          <PanelText
+            id={`${title}-${panelId}`}
+            aria-label={sprintf(
+              /* translators: %d: number of issues, %s: title */
```
Comment (swissspidy) [inline assets/src/edit-story/components/tablist/tablistPanel.js]:
```suggestion
              /* translators: 1: number of issues. 2: type of issue, for example "High Priority". */
```

**Review by BrittanyIRL — APPROVED** [review]
beautiful, thank you again for finishing this before you went on vacation.

File: assets/src/edit-story/components/tablist/styles.js (hunk 12 above)
Comment (maxyinger) [reply assets/src/edit-story/components/tablist/styles.js:98]: I did a small change here too. might cause some conflicts, but just gotta keep track of which merges first and update the other:
https://github.com/google/web-stories-wp/pull/8225

File: assets/src/edit-story/masks/display.js (hunk 8 above)
Comment (maxyinger) [reply assets/src/edit-story/masks/display.js:86]: 🔥

**Review by maxyinger — APPROVED** [review]
This all looks great to me. Good work!

**Pull request merged**